ownCloud/Viewer/PDFThumbnailsCollectionViewController.swift

```diff
@@ -0,0 +1,203 @@
+        flowLayout?.sectionInset = UIEdgeInsets(top: verticalInset, left: horizontalInset, bottom: verticalInset, right: horizontalInset)
+    }
+
+    // MARK: - Themeable support
+
+    func applyThemeCollection(theme: Theme, collection: ThemeCollection, event: ThemeEvent) {
+        self.collectionView!.applyThemeCollection(collection)
+    }
+
+    // MARK: UICollectionViewDataSource
+
+    override func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return 1
+    }
+
+    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return self.pdfDocument?.pageCount ?? 0
+    }
+
+    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: PDFThumbnailCollectionViewCell.identifier, for: indexPath) as? PDFThumbnailCollectionViewCell
+        let (image, label) =  getCachedThumbnailOrFetchAsync(at: indexPath)
+        cell?.imageView?.image = image
+        cell?.pageLabel?.text = label
+        return cell!
+    }
+
+    // MARK: - UICollectionViewDelegate
+
+    override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        if let pdfPage = self.pdfDocument?.page(at: indexPath.item) {
+            self.dismiss(animated: true) {
+                NotificationCenter.default.post(name: PDFViewerViewController.PDFGoToPageNotification.name, object: pdfPage)
+            }
+        }
+    }
+
+    override func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
+        if !decelerate {
+            updateVisibleCells()
+        }
+    }
+
+    override func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        updateVisibleCells()
+    }
+
+    // МАРК: - UICollectionViewDataSourcePrefetching
+
+    func collectionView(_ collectionView: UICollectionView, prefetchItemsAt indexPaths: [IndexPath]) {
+
+        for indexPath in indexPaths {
+            if let pdfPage = self.pdfDocument?.page(at: indexPath.item) {
+                if let layout = self.collectionView?.collectionViewLayout as? UICollectionViewFlowLayout {
+                    fetchThumbnail(for: pdfPage, indexPath: indexPath, size: layout.itemSize)
+                }
+            }
+        }
+    }
+
+    func collectionView(_ collectionView: UICollectionView, cancelPrefetchingForItemsAt indexPaths: [IndexPath]) {
+
+        for indexPath in indexPaths {
+            cancelFetchingThumbnail(indexPath: indexPath)
+        }
+    }
+
+    // MARK: - Private methods
+
+    fileprivate func fetchThumbnail(for page:PDFPage, indexPath:IndexPath, size:CGSize) {
+        let blockOperation = BlockOperation()
+        weak var weakBlockOperation = blockOperation
+
+        blockOperation.addExecutionBlock { [unowned self] in
```

Comment:
If the `PDFThumbnailsCollectionViewController` is dismissed and deallocated while a thumbnail is being fetched, the app will crash because the reference to self is `unowned`. Using `[weak self]` instead would account for that case.